drivers/timer/amd_tmrctr.c

```diff
@@ -0,0 +1,291 @@
+	mem_addr_t base;
+	uint32_t clock_rate;
+	uint32_t irq;
+	bool one_timer;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(interrupts)
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+/* private data */
+struct xilinx_timer_data {
+	uint32_t clocksource_offset;
+	uint32_t clockevent_offset;
+};
+
+/* Pointing to timer instance which is system timer */
+static const struct device *sys_dev;
+
+static inline uint32_t xlnx_tmrctr_read32(const struct device *dev, uint32_t timer_offset,
+					  uint32_t offset)
+{
+	const struct xilinx_timer_config *config = dev->config;
+	uint32_t reg = (uint32_t)(config->base) + timer_offset + offset;
+
+	LOG_DBG("%s: 0x%x (base = 0x%lx, timer_offset = 0x%x, offset = 0x%x",
+		__func__, reg, config->base, timer_offset, offset);
+
+	return sys_read32(reg);
+}
+
+static inline void xlnx_tmrctr_write32(const struct device *dev, uint32_t timer_offset,
+				       uint32_t value, uint32_t offset)
+{
+	const struct xilinx_timer_config *config = dev->config;
+	uint32_t reg = (uint32_t)(config->base) + timer_offset + offset;
+
+	LOG_DBG("%s: 0x%x (base = 0x%lx, timer_offset = 0x%x, offset = 0x%x",
+		__func__, reg, config->base, timer_offset, offset);
+
+	sys_write32(value, reg);
+}
+
+volatile uint32_t xlnx_tmrctr_read_count(const struct device *dev)
+{
+	struct xilinx_timer_data *data = dev->data;
+
+	return xlnx_tmrctr_read32(dev, data->clocksource_offset, XTC_TCR_OFFSET);
+}
+
+volatile uint32_t xlnx_tmrctr_read_hw_cycle_count(const struct device *dev)
+{
+	const struct xilinx_timer_config *config = dev->config;
+
+	return (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / config->clock_rate) *
+				 xlnx_tmrctr_read_count(dev);
+}
+
+static void xlnx_tmrctr_clear_interrupt(const struct device *dev)
+{
+	struct xilinx_timer_data *data = dev->data;
+
+	uint32_t control_status_register = xlnx_tmrctr_read32(dev, data->clockevent_offset,
+							      XTC_TCSR_OFFSET);
+
+	xlnx_tmrctr_write32(dev, data->clockevent_offset,
+			    control_status_register | XTC_CSR_INT_OCCURRED_MASK, XTC_TCSR_OFFSET);
+}
+
+static void xlnx_tmrctr_irq_handler(const struct device *dev)
+{
+	uint32_t delta_ticks;
+	const struct xilinx_timer_config *config = dev->config;
+	uint32_t cycles = xlnx_tmrctr_read_hw_cycle_count(dev);
+
+	LOG_DBG("%s: Base 0x%lx IRQ happened\n", __func__, config->base);
+
+	/*
+	 * TODO: CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is likely all the time equal to timer
+	 * frequency that's why pretty much all calculation around it should be fixed.
+	 */
+	delta_ticks = (cycles - last_cycles) / CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
+	if (delta_ticks == 0) {
+		/*
+		 * This is an implicit TICKFUL mode implementation
+		 * We expect the cycle difference to be very low and
+		 * hence the delta_ticks to be 0, if IRQ fires on a timely basis.
+		 * But if an IRQ fire was missed due to some long running critical
+		 * section or TICKLESS mode, then delta should be large enough.
+		 */
+		delta_ticks = 1;
+	}
+	last_cycles = cycles;
```

Comment:
One more nitpick.

This is a slippage bug here: "cycles" contains the actual real time you read above, and you will announce a number of ticks based on the delta between that and the time of the last interrupt.  But let's say you're (due to interrupt latency or whatever) 7 cycles later than you expected. Well, the **next** set of ticks will be computed based on a "last tick" that was seven cycles later than expected, and so on.  And the end result is that your clock is going to count somewhat slower than real time.

The right way to do this is to leave last_cycles in tick granularity, storing not the time the last interrupt was received, but when it was scheduled for.  You can look at the other drivers for guides.  The rounding and cycle/tick math for timer drivers does get a little subtle.

@npitre has been in this world most recently and might have more specific advice.

Reply:
Hi, original author here. This is a slightly outdated implementation where I've noticed and fixed this (too late clearly). On the bright side of things, happy to report `timer` tests are doing a great job!

https://github.com/zephyrproject-rtos/zephyr/pull/53576/files#diff-69c26924180ecb97861e90a01fed45b2884fe337c155b1f834d853090c9cdcdf

I'll sync with @michalsimek .

Reply:
I have pushed new version which should be aligned with apic_tsc.c code. Please check.